File: dart/src/sql_formatter.dart
class T {
  String str;
  List<String> shiftArr;
  String tab;
  List<String> arr;
  int parensLevel;
  int deep;

  T(String str, List<String> shiftArr, String tab, List<String> arr, int parensLevel, int deep) {
    this.str = str;
    this.shiftArr = shiftArr;
    this.tab = tab;
    this.arr = arr;
    this.parensLevel = parensLevel;
    this.deep = deep;
  }
}

class Out {
  String str;
  int deep;

  Out(String str, int deep) {
    this.str = str;
    this.deep = deep;
  }
}

class Replacement {
  RegExp pattern;
  String replacement;

  Replacement(RegExp pattern, String replacement) {
    this.pattern = pattern;
    this.replacement = replacement;
  }
}

class SQLFormatter {
  static const sep = '~::~';

  static String format(String sql, int numSpaces) {
    String tab = ' ' * numSpaces;
    List<String> splitByQuotes = transformString(sql, [
      (s) => s.replaceAll(new RegExp(r'\s+'), ' '),
      (s) => s.replaceAll("'", "$sep'")
    ]).split(sep);
    T input = new T('', createShiftArr(tab), tab, genArray(splitByQuotes, tab), 0, 0);
    T output = new List<int>.generate(input.arr.length, (int i) => i).fold(input, updateOutput);

    return transformString(output.str, [
      (s) => s.replaceAll(new RegExp(r'\s+\n'), '\n'),
      (s) => s.replaceAll(new RegExp(r'\n+'), '\n'),
      (s) => s.trim()
    ]);
  }

  static T updateOutput(T acc, int i) {
    String originalEl = acc.arr[i];
    int parensLevel = subqueryLevel(originalEl, acc.parensLevel);
    List<String> arr = new List<String>.from(acc.arr);
    if (originalEl.contains(new RegExp(r'SELECT|SET'))) {
      arr[i] = originalEl.replaceAll(new RegExp(r'\,\s+'), ',\n${acc.tab}${acc.tab}');
    }
    String el = arr[i];
    Out out = el.contains(new RegExp(r'\(\s*SELECT'))
      ? new Out('${acc.str}${acc.shiftArr[acc.deep + 1]}${el}', acc.deep + 1)
      : new Out(
        el.contains("'") ? '${acc.str}${el}' : '${acc.str}${acc.shiftArr[acc.deep]}${el}',
        parensLevel < 1 && acc.deep != 0 ? acc.deep - 1 : acc.deep
      );

    return new T(out.str, acc.shiftArr, acc.tab, arr, parensLevel, out.deep);
  }

  static String transformString(String str, List<Function> fns) => fns.fold(str, (acc, fn) => fn(acc));

  static List<String> createShiftArr(String tab) => new List<String>.generate(100, (int i) => '\n' + (tab * i));
  static List<String> genArray(List<String> splitByQuotes, String tab) {
    return new List<List<String>>.generate(
      splitByQuotes.length,
      (int i) => splitIfEven(i, splitByQuotes[i], tab)
    ).expand((x) => x).toList();
  }

  static int subqueryLevel(String str, int level) {
    return level - (str.replaceAll('(', '').length - str.replaceAll(')', '').length);
  }

  static List<Replacement> allReplacements(String tab) {
    return [
      new Replacement(new RegExp(r' AND ', caseSensitive: false),              sep + tab + 'AND '),
      new Replacement(new RegExp(r' BETWEEN ', caseSensitive: false),          sep + tab + 'BETWEEN '),
      new Replacement(new RegExp(r' CASE ', caseSensitive: false),             sep + tab + 'CASE '),
      new Replacement(new RegExp(r' ELSE ', caseSensitive: false),             sep + tab + 'ELSE '),
      new Replacement(new RegExp(r' END ', caseSensitive: false),              sep + tab + 'END '),
      new Replacement(new RegExp(r' FROM ', caseSensitive: false),             sep + 'FROM '),
      new Replacement(new RegExp(r' GROUP\s+BY ', caseSensitive: false),       sep + 'GROUP BY '),
      new Replacement(new RegExp(r' HAVING ', caseSensitive: false),           sep + 'HAVING '),
      new Replacement(new RegExp(r' IN ', caseSensitive: false),               ' IN '),
      new Replacement(new RegExp(r' JOIN ', caseSensitive: false),             sep + 'JOIN '),
      new Replacement(new RegExp(r' CROSS(~::~)+JOIN ', caseSensitive: false), sep + 'CROSS JOIN '),
      new Replacement(new RegExp(r' INNER(~::~)+JOIN ', caseSensitive: false), sep + 'INNER JOIN '),
      new Replacement(new RegExp(r' LEFT(~::~)+JOIN ', caseSensitive: false),  sep + 'LEFT JOIN '),
      new Replacement(new RegExp(r' RIGHT(~::~)+JOIN ', caseSensitive: false), sep + 'RIGHT JOIN '),
      new Replacement(new RegExp(r' ON ', caseSensitive: false),               sep + tab + 'ON '),
      new Replacement(new RegExp(r' OR ', caseSensitive: false),               sep + tab + 'OR '),
      new Replacement(new RegExp(r' ORDER\s+BY ', caseSensitive: false),       sep + 'ORDER BY '),
      new Replacement(new RegExp(r' OVER ', caseSensitive: false),             sep + tab + 'OVER '),
      new Replacement(new RegExp(r'\(\s*SELECT ', caseSensitive: false),       sep + '(SELECT '),
      new Replacement(new RegExp(r'\)\s*SELECT ', caseSensitive: false),       ')' + sep + 'SELECT '),
      new Replacement(new RegExp(r' THEN ', caseSensitive: false),             ' THEN' + sep + tab),
      new Replacement(new RegExp(r' UNION ', caseSensitive: false),            sep + 'UNION' + sep),
      new Replacement(new RegExp(r' USING ', caseSensitive: false),            sep + 'USING '),
      new Replacement(new RegExp(r' WHEN ', caseSensitive: false),             sep + tab + 'WHEN '),
      new Replacement(new RegExp(r' WHERE ', caseSensitive: false),            sep + 'WHERE '),
      new Replacement(new RegExp(r' WITH ', caseSensitive: false),             sep + 'WITH '),
      new Replacement(new RegExp(r' SET ', caseSensitive: false),              sep + 'SET '),
      new Replacement(new RegExp(r' ALL ', caseSensitive: false),              ' ALL '),
      new Replacement(new RegExp(r' AS ', caseSensitive: false),               ' AS '),
      new Replacement(new RegExp(r' ASC ', caseSensitive: false),              ' ASC '),
      new Replacement(new RegExp(r' DESC ', caseSensitive: false),             ' DESC '),
      new Replacement(new RegExp(r' DISTINCT ', caseSensitive: false),         ' DISTINCT '),
      new Replacement(new RegExp(r' EXISTS ', caseSensitive: false),           ' EXISTS '),
      new Replacement(new RegExp(r' NOT ', caseSensitive: false),              ' NOT '),
      new Replacement(new RegExp(r' NULL ', caseSensitive: false),             ' NULL '),
      new Replacement(new RegExp(r' LIKE ', caseSensitive: false),             ' LIKE '),
      new Replacement(new RegExp(r'\s*SELECT ', caseSensitive: false),         'SELECT '),
      new Replacement(new RegExp(r'\s*UPDATE ', caseSensitive: false),         'UPDATE '),
      new Replacement(new RegExp(r'\s*DELETE ', caseSensitive: false),         'DELETE '),
      new Replacement(new RegExp('($sep)+'),                                      sep)
    ];
  }

  static List<String> splitSql(String str, String tab) {
    return allReplacements(tab).fold(str.replaceAll(new RegExp(r'\s+'), ' '), (String acc, Replacement r) {
      return acc.replaceAll(r.pattern, r.replacement);
    }).split(sep);
  }

  static List<String> splitIfEven(int i, String str, String tab) => i % 2 == 0 ? splitSql(str, tab) : [str];
}

File: php/src/SQLFormatter.php
<?php

class SQLFormatter {
  public $sep = '~::~';

  public function format($sql, $numSpaces) {
    $tab = str_repeat(' ', $numSpaces);
    $splitByQuotes = explode($this->sep, preg_replace('/\s+/', ' ', preg_replace("/'/", $this->sep . "'", $sql)));
    $input = array(
      'str' => '',
      'shiftArr' => $this->createShiftArr($tab),
      'tab' => $tab,
      'parensLevel' => 0,
      'deep' => 0
    );

    $inArr = $this->genArray($splitByQuotes, $tab);
    $outArr = array();
    $acc = $input;
    foreach($inArr as $i => $originalEl) {
      $parensLevel = $this->subqueryLevel($originalEl, $acc['parensLevel']);
      $el = preg_match('/SELECT|SET/', $originalEl) ? preg_replace('/,\s+/', ",\n" . $acc['tab'] . $acc['tab'], $originalEl) : $originalEl;
      $out = $this->updateOutput($el, $parensLevel, $acc);

      $outArr[$i] = $el;
      $acc['str'] = $out[0];
      $acc['parensLevel'] = $parensLevel;
      $acc['deep'] = $out[1];
    }

    return trim(preg_replace('/\s+\\n/', "\n", preg_replace('/\\n+/', "\n", $acc['str'])));
  }

  private function updateOutput($el, $parensLevel, $acc) {
    if (preg_match('/\(\s*SELECT/', $el)) {
      return array(($acc['str'] . $acc['shiftArr'][$acc['deep'] + 1] . $el), $acc['deep'] + 1);
    } else {
      return array(
        (preg_match("/'/", $el) ? ($acc['str'] . $el) : ($acc['str'] . $acc['shiftArr'][$acc['deep']] . $el)),
        ($parensLevel < 1 && $acc['deep'] !== 0 ? $acc['deep'] - 1 : $acc['deep'])
      );
    }
  }

  private function createShiftArr($tab) {
    $arr = array();
    for ($i = 0; $i < 100; $i++) {
      array_push($arr, "\n" . str_repeat($tab, $i));
    }
    return $arr;
  }

  private function genArray($splitByQuotes, $tab) {
    $arr = array();
    foreach($splitByQuotes as $i => $a) {
      foreach($this->splitIfEven($i, $splitByQuotes[$i], $tab) as $el) {
        $arr[] = $el;
      }
    }
    return $arr;
  }

  private function subqueryLevel($str, $level) {
    return $level - (strlen(preg_replace('/\(/', '', $str)) - strlen(preg_replace('/\)/', '', $str)));
  }

  private function allReplacements($tab) {
    $sep = $this->sep;
    return array(
      function($str) use($tab, $sep) { return preg_replace('/ AND /i',              $sep . $tab . 'AND ', $str); },
      function($str) use($tab, $sep) { return preg_replace('/ BETWEEN /i',          $sep . $tab . 'BETWEEN ', $str); },
      function($str) use($tab, $sep) { return preg_replace('/ CASE /i',             $sep . $tab . 'CASE ', $str); },
      function($str) use($tab, $sep) { return preg_replace('/ ELSE /i',             $sep . $tab . 'ELSE ', $str); },
      function($str) use($tab, $sep) { return preg_replace('/ END /i',              $sep . $tab . 'END ', $str); },
      function($str) use($tab, $sep) { return preg_replace('/ FROM /i',             $sep . 'FROM ', $str); },
      function($str) use($tab, $sep) { return preg_replace('/ GROUP\s+BY /i',       $sep . 'GROUP BY ', $str); },
      function($str) use($tab, $sep) { return preg_replace('/ HAVING /i',           $sep . 'HAVING ', $str); },
      function($str) use($tab, $sep) { return preg_replace('/ IN /i',               ' IN ', $str); },
      function($str) use($tab, $sep) { return preg_replace('/ JOIN /i',             $sep . 'JOIN ', $str); },
      function($str) use($tab, $sep) { return preg_replace('/ CROSS(~::~)+JOIN /i', $sep . 'CROSS JOIN ', $str); },
      function($str) use($tab, $sep) { return preg_replace('/ INNER(~::~)+JOIN /i', $sep . 'INNER JOIN ', $str); },
      function($str) use($tab, $sep) { return preg_replace('/ LEFT(~::~)+JOIN /i',  $sep . 'LEFT JOIN ', $str); },
      function($str) use($tab, $sep) { return preg_replace('/ RIGHT(~::~)+JOIN /i', $sep . 'RIGHT JOIN ', $str); },
      function($str) use($tab, $sep) { return preg_replace('/ ON /i',               $sep . $tab . 'ON ', $str); },
      function($str) use($tab, $sep) { return preg_replace('/ OR /i',               $sep . $tab . 'OR ', $str); },
      function($str) use($tab, $sep) { return preg_replace('/ ORDER\s+BY /i',       $sep . 'ORDER BY ', $str); },
      function($str) use($tab, $sep) { return preg_replace('/ OVER /i',             $sep . $tab . 'OVER ', $str); },
      function($str) use($tab, $sep) { return preg_replace('/\(\s*SELECT /i',       $sep . '(SELECT ', $str); },
      function($str) use($tab, $sep) { return preg_replace('/\)\s*SELECT /i',       ')' . $sep . 'SELECT ', $str); },
      function($str) use($tab, $sep) { return preg_replace('/ THEN /i',             ' THEN' . $sep . $tab, $str); },
      function($str) use($tab, $sep) { return preg_replace('/ UNION /i',            $sep . 'UNION' . $sep, $str); },
      function($str) use($tab, $sep) { return preg_replace('/ USING /i',            $sep . 'USING ', $str); },
      function($str) use($tab, $sep) { return preg_replace('/ WHEN /i',             $sep . $tab . 'WHEN ', $str); },
      function($str) use($tab, $sep) { return preg_replace('/ WHERE /i',            $sep . 'WHERE ', $str); },
      function($str) use($tab, $sep) { return preg_replace('/ WITH /i',             $sep . 'WITH ', $str); },
      function($str) use($tab, $sep) { return preg_replace('/ SET /i',              $sep . 'SET ', $str); },
      function($str) use($tab, $sep) { return preg_replace('/ ALL /i',              ' ALL ', $str); },
      function($str) use($tab, $sep) { return preg_replace('/ AS /i',               ' AS ', $str); },
      function($str) use($tab, $sep) { return preg_replace('/ ASC /i',              ' ASC ', $str); },
      function($str) use($tab, $sep) { return preg_replace('/ DESC /i',             ' DESC ', $str); },
      function($str) use($tab, $sep) { return preg_replace('/ DISTINCT /i',         ' DISTINCT ', $str); },
      function($str) use($tab, $sep) { return preg_replace('/ EXISTS /i',           ' EXISTS ', $str); },
      function($str) use($tab, $sep) { return preg_replace('/ NOT /i',              ' NOT ', $str); },
      function($str) use($tab, $sep) { return preg_replace('/ NULL /i',             ' NULL ', $str); },
      function($str) use($tab, $sep) { return preg_replace('/ LIKE /i',             ' LIKE ', $str); },
      function($str) use($tab, $sep) { return preg_replace('/\s*SELECT /i',         'SELECT ', $str); },
      function($str) use($tab, $sep) { return preg_replace('/\s*UPDATE /i',         'UPDATE ', $str); },
      function($str) use($tab, $sep) { return preg_replace('/\s*DELETE /i',         'DELETE ', $str); },
      function($str) use($tab, $sep) { return preg_replace('/(~::~)+/', $sep, $str); }
    );
  }

  private function splitSql($str, $tab) {
    $str = preg_replace('/\s+/', ' ', $str);
    $replacements = $this->allReplacements($tab);
    foreach($replacements as $replacementFn) {
      $str = $replacementFn($str);
    }
    return explode($this->sep, $str);
  }

  private function splitIfEven($i, $str, $tab) {
    return ($i % 2) === 0 ? $this->splitSql($str, $tab) : array($str);
  }
}

if ($_ENV['EXPORTS'] !== 'false') {
  $module->exports = new SQLFormatter;
}
